File: frontend/mobile/lib/core/models/device_model.dart
import 'package:json_annotation/json_annotation.dart';

part 'device_model.g.dart';

enum DeviceStatus {
  @JsonValue('Active')
  active,
  @JsonValue('Stolen')
  stolen,
  @JsonValue('Recovered')
  recovered,
  @JsonValue('ForSale')
  forSale,
  @JsonValue('Inactive')
  inactive,
}

@JsonSerializable()
class Device {
  final String id;
  final String serialNumber;
  final String category;
  final String? brand;
  final String? model;
  final String? description;
  final DeviceStatus status;
  final String userId;
  final String verificationCode;
  final bool isVerified;
  final DateTime registeredAt;
  final DateTime lastUpdatedAt;

  Device({
    required this.id,
    required this.serialNumber,
    required this.category,
    this.brand,
    this.model,
    this.description,
    required this.status,
    required this.userId,
    required this.verificationCode,
    required this.isVerified,
    required this.registeredAt,
    required this.lastUpdatedAt,
  });

  factory Device.fromJson(Map<String, dynamic> json) => _$DeviceFromJson(json);
  Map<String, dynamic> toJson() => _$DeviceToJson(this);

  String get displayName => '${brand ?? ''} ${model ?? category}'.trim();
}

@JsonSerializable()
class RegisterDeviceRequest {
  final String serialNumber;
  final String category;
  final String? brand;
  final String? model;
  final String? description;
  final DateTime? purchaseDate;
  final double? purchasePrice;
  final String? retailer;

  RegisterDeviceRequest({
    required this.serialNumber,
    required this.category,
    this.brand,
    this.model,
    this.description,
    this.purchaseDate,
    this.purchasePrice,
    this.retailer,
  });

  Map<String, dynamic> toJson() => _$RegisterDeviceRequestToJson(this);
}

File: frontend/web/src/app/core/models/device.model.ts
export enum DeviceStatus {
  Active = 'Active',
  Stolen = 'Stolen',
  Recovered = 'Recovered',
  ForSale = 'ForSale',
  Inactive = 'Inactive'
}

export interface Device {
  id: string;
  serialNumber: string;
  category: string;
  brand?: string;
  model?: string;
  description?: string;
  status: DeviceStatus;
  userId: string;
  verificationCode: string;
  isVerified: boolean;
  registeredAt: Date;
  lastUpdatedAt: Date;
  photos?: DevicePhoto[];
  documents?: DeviceDocument[];
}

export interface DevicePhoto {
  id: string;
  deviceId: string;
  photoUrl: string;
  caption?: string;
  isPrimary: boolean;
  uploadedAt: Date;
}

export interface DeviceDocument {
  id: string;
  deviceId: string;
  documentType: string;
  documentUrl: string;
  fileName: string;
  uploadedAt: Date;
}

export interface RegisterDeviceRequest {
  serialNumber: string;
  category: string;
  brand?: string;
  model?: string;
  description?: string;
  purchaseDate?: Date;
  purchasePrice?: number;
  retailer?: string;
}

export interface DeviceResponse {
  id: string;
  serialNumber: string;
  category: string;
  brand?: string;
  model?: string;
  description?: string;
  status: string;
  verificationCode: string;
  isVerified: boolean;
  registeredAt: Date;
}
